File: cli/src/modules/dockerfile/ai-clis.ts
import { POST_SCRIPT_DIR, type DockerfileModule } from '../../types.js';

type CliId =
  | 'claude-code'
  | 'opencode'
  | 'autoskills'
  | 'antigravity-cli'
  | 'copilot-cli';

const ALL: CliId[] = [
  'claude-code',
  'opencode',
  'autoskills',
  'antigravity-cli',
  'copilot-cli',
];

const SCRIPT_BY_TOOL: Record<CliId, string> = {
  'claude-code': 'install-claude-code.sh',
  opencode: 'install-opencode.sh',
  autoskills: 'install-autoskills.sh',
  'antigravity-cli': 'install-antigravity.sh',
  'copilot-cli': 'install-copilot.sh',
};

export const aiClisModule: DockerfileModule = {
  id: 'ai-clis',
  label: 'AI CLIs install scripts (Claude Code, OpenCode, Autoskills, Antigravity, Copilot CLI)',
  category: 'infra',
  requires: ['pnpm', 'github-cli'],
  // Shipped as post-install scripts: baked into POST_SCRIPT_DIR by the generator,
  // available inside the container, not auto-run.
  postScriptFiles: (opts) => normalizeTools(opts.tools).map((t) => SCRIPT_BY_TOOL[t]),
  options: [
    {
      id: 'tools',
      label: 'AI CLIs to ship install scripts for',
      type: 'multiselect',
      choices: [
        { value: 'claude-code', label: 'Claude Code (native standalone installer)' },
        { value: 'opencode', label: 'OpenCode (opencode.ai installer)' },
        { value: 'autoskills', label: 'Autoskills (npx, no global install)' },
        { value: 'antigravity-cli', label: 'Antigravity CLI (native standalone installer)' },
        { value: 'copilot-cli', label: 'GitHub Copilot CLI (standalone binary)' },
      ],
      default: ALL,
    },
  ],
  render() {
    return `##\n## AI CLIs — install scripts shipped under ${POST_SCRIPT_DIR}\n##\n`;
  },
};

function normalizeTools(raw: unknown): CliId[] {
  if (!Array.isArray(raw)) return ALL;
  const set = new Set(raw.map(String));
  return ALL.filter((t) => set.has(t));
}

File: cli-go/internal/modules/dockerfile/ai_clis.go
package dockerfile

import (
	"fmt"

	"github.com/joacohbc/my-devcontainer-installer/cli-go/internal/core"
)

var allAiCliIDs = []string{
	"claude-code",
	"opencode",
	"codex-cli",
	"antigravity-cli",
	"copilot-cli",
}

var scriptByTool = map[string]string{
	"claude-code":     "install-claude-code.sh",
	"opencode":        "install-opencode.sh",
	"codex-cli":       "install-codex-cli.sh",
	"antigravity-cli": "install-antigravity.sh",
	"copilot-cli":     "install-copilot.sh",
}

var AiClisModule = &DockerfileModuleSpec{
	ID:       "ai-clis",
	Label:    "AI CLIs install scripts (Claude Code, OpenCode, Codex CLI, Antigravity, Copilot CLI)",
	Category: core.CategoryInfra,
	Requires: []string{"pnpm", "github-cli"},
	Options: []core.ModuleOption{
		{
			ID:    "tools",
			Label: "AI CLIs to ship install scripts for",
			Type:  core.ModuleOptionMultiselect,
			Choices: []core.ModuleOptionChoice{
				{Value: "claude-code", Label: "Claude Code (native standalone installer)"},
				{Value: "opencode", Label: "OpenCode (opencode.ai installer)"},
				{Value: "codex-cli", Label: "Codex CLI (npx, no global install)"},
				{Value: "antigravity-cli", Label: "Antigravity CLI (native standalone installer)"},
				{Value: "copilot-cli", Label: "GitHub Copilot CLI (standalone binary)"},
			},
			Default: allAiCliIDs,
		},
	},
	PostScriptFiles: func(opts map[string]any) []string {
		tools := normalizeAiTools(opts["tools"])
		scripts := make([]string, 0, len(tools))
		for _, t := range tools {
			if script, ok := scriptByTool[t]; ok {
				scripts = append(scripts, script)
			}
		}
		return scripts
	},
	Render: func(opts map[string]any) string {
		return fmt.Sprintf("##\n## AI CLIs — install scripts shipped under %s\n##\n", core.PostScriptDir)
	},
}

func normalizeAiTools(raw any) []string {
	arr, ok := raw.([]any)
	if !ok {
		return allAiCliIDs
	}
	rawSet := make(map[string]bool, len(arr))
	for _, v := range arr {
		if s, ok := v.(string); ok {
			rawSet[s] = true
		}
	}
	result := make([]string, 0, len(arr))
	for _, id := range allAiCliIDs {
		if rawSet[id] {
			result = append(result, id)
		}
	}
	return result
}
